grid2op/Rules/rulesByArea.py

```diff
@@ -0,0 +1,119 @@
+# SPDX-License-Identifier: MPL-2.0
+# This file is part of Grid2Op, Grid2Op a testbed platform to model sequential decision making in power systems.
+
+import numpy as np
+from itertools import chain
+from grid2op.Rules.BaseRules import BaseRules
+from grid2op.Rules.LookParam import LookParam
+from grid2op.Rules.PreventReconnection import PreventReconnection
+from grid2op.Rules.PreventDiscoStorageModif import PreventDiscoStorageModif
+from grid2op.Exceptions import (
+    IllegalAction,
+)
+
+class RulesByArea(BaseRules):
+    """
+    This subclass combine :class:`PreventReconnection`, :class: `PreventDiscoStorageModif` to be applied on the whole grid at once,
+    while a specifique method look for the legality of simultaneous actions taken on defined areas of a grid.
+    An action is declared legal if and only if:
+
+      - It doesn't reconnect more power lines than what is stated in the actual game _parameters
+        :class:`grid2op.Parameters`
+      - It doesn't attempt to act on more substations and lines within each area that what is stated in the actual game _parameters
+        :class:`grid2op.Parameters`
+      - It doesn't attempt to modify the power produce by a turned off storage unit
+      
+    Example
+    ---------
+    If you want the environment to take into account the rules by area, you can achieve it with:
+
+    .. code-block:
+        import grid2op
+        from grid2op.Rules.rulesByArea import RulesByArea
+        
+        # First you set up the areas within the RulesByArea class
+        my_gamerules_byarea = RulesByArea([[0,1,2,3,4,5,6,7],[8,9,10,11,12,13,14]])
+        # Then you create your environment with it:
+        NAME_OF_THE_ENVIRONMENT = "rte_case14_realistic"
+        env = grid2op.make(NAME_OF_THE_ENVIRONMENT,gamerules_class=my_gamerules_byarea)
+
+    """
+
+    def __init__(self, areas_list):
+        """
+        The initialization of the rule with a list of list of ids of substations composing the aimed areas.
+        Parameters
+        ----------
+        areas_list : list of areas, each placeholder containing the ids of substations of each defined area
+        """
+        self.substations_id_by_area = {i : sorted(k) for i,k in enumerate(areas_list)}
+        
+        
+    def initialize(self, env):
+        """
+        This function is used to inform the class instance about the environment specification and check no substation of the grid are left ouside an area. 
+        Parameters
+        ----------
+        env: :class:`grid2op.Environment.Environment`
+            An environment instance properly initialized.
+
+        """
+        n_sub = env.n_sub
+        n_sub_rule = np.sum([len(set(list_ids)) for list_ids in self.substations_id_by_area.values()])
+        if n_sub_rule != n_sub: 
+            print("The number of listed ids of substations in rule initialization does not match the number of substations of the chosen environement. Look for missing ids or doublon")
```

Comment:
can you raise a Grid2opException here instead ?

Reply:
ok !